### PR TITLE
Rewrite some calculation functions

### LINKED ISSUE #221: Small refactoring
Simplify some calculations

### PR DESCRIPTION
**Pull request by IraSoro**
Closed #221 

I rewrote several calculation functions. The logic of most of the calculations remained the same, I just rewrote it more beautifully. And also added comments.

All functions have passed the tests, but I still want you to check everything as carefully as possible.

### CODE REVIEW

File: src/pages/TabHome.tsx [hunk 1]
```diff
@@ -96,9 +96,9 @@ const ViewCalendar = (props: SelectCalendarProps) => {
   const { cycles } = useContext(CyclesContext);
   const theme = useContext(ThemeContext).theme;
 
-  const periodDays = getPeriodDays(cycles);
-  const forecastPeriodDays = getForecastPeriodDays(cycles);
-  const ovulationDays = getOvulationDays(cycles);
+  const periodDays = getPeriodDates(cycles);
```
Comment (imblowfish) [inline src/pages/TabHome.tsx:99]: const **periodDates** = getPeriodDates(cycles);

File: src/state/CalculationLogics.ts [hunk 2]
```diff
@@ -26,53 +26,69 @@ export function getLengthOfLastPeriod(cycles: Cycle[]) {
...
+  return differenceInDays(currentDate, start) + 1;
 }
 
-export function getPregnancyChance(cycles: Cycle[]) {
-  if (cycles.length === 0) {
-    return "";
-  }
+// NOTE: Detailed description of ovulation calculation: https://github.com/IraSoro/peri/blob/master/info/CALCULATION.md#ovulation-day
+export function getOvulationStatus(cycles: Cycle[]) {
+  if (cycles.length === 0) return "";
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:45]: Please add `{` and `}` here

If we'll try to debug this function in the future then we should add `{` `}` every time ourselves, why don't add them right away?

File: src/state/CalculationLogics.ts [hunk 3]
```diff
@@ -26,53 +26,69 @@ export function getLengthOfLastPeriod(cycles: Cycle[]) {
...
+  // Calculate what day ovulation should occur based on cycle length
+  // For example, if the cycle length is 30, then ovulation should occur on 30-14=16 day
   const ovulationDay = cycleLength - lutealPhaseLength;
+  // Calculate the difference between the current day of the cycle and the day of ovulation
   const diffDay = ovulationDay - dayOfCycle;
 
-  if (diffDay >= -2 && diffDay <= 1) {
+  // If the day of ovulation has already passed by more than 2 days, return the status "completed"
+  // NOTE: 2 is the error value (margin of error)
+  if (diffDay < -2) return i18n.t("finished");
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:60]: Please add `{` and `}` here

File: src/state/CalculationLogics.ts [hunk 4]
```diff
@@ -26,53 +26,69 @@ export function getLengthOfLastPeriod(cycles: Cycle[]) {
...
+
+  switch (diffDay) {
+    case -2:
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:63]:
```ts
switch (diffDay) {
  case -2:
  case -1:
    return i18n.t("possible");
}
```

File: src/state/CalculationLogics.ts [hunk 5]
```diff
@@ -26,53 +26,69 @@ export function getLengthOfLastPeriod(cycles: Cycle[]) {
...
+    default:
+      return `${i18n.t("in")} ${diffDay} ${i18n.t("Days", {
+        postProcess: "interval",
+        count: diffDay,
+      })}`;
+  }
+}
+
+export function getPregnancyChance(cycles: Cycle[]) {
+  if (cycles.length === 0) return "";
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:81]: `{` `}`

File: src/state/CalculationLogics.ts [hunk 6]
```diff
@@ -300,42 +317,43 @@ export function getNewCyclesHistory(periodDays: string[]) {
...
 }
 
-export function getLastPeriodDays(cycles: Cycle[]) {
-  const periodDays: string[] = [];
-  const lastCycle = cycles.at(0);
-  if (!lastCycle) {
-    return periodDays;
-  }
+export function getPeriodDatesOfLastCycle(cycles: Cycle[]) {
+  if (cycles.length === 0) return [];
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:346]: `{` `}`

File: src/state/CalculationLogics.ts [hunk 7]
```diff
@@ -352,72 +370,82 @@ export function getActiveDates(date: Date, cycles: Cycle[]) {
...
-  return periodDates;
+  return periodDates.concat(additionalDays);
 }
 
-export function getForecastPeriodDays(cycles: Cycle[]) {
-  if (cycles.length === 0) {
-    return [];
-  }
+export function getForecastPeriodDates(cycles: Cycle[]) {
+  if (cycles.length === 0) return [];
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:408]: `{` `}`

File: src/state/CalculationLogics.ts [hunk 8]
```diff
@@ -352,72 +370,82 @@ export function getActiveDates(date: Date, cycles: Cycle[]) {
...
+
   addForecastDates(nextCycleStart);
-  if (cycles.length === 1) {
-    return forecastDates;
-  }
 
-  const cycleCount = 6;
-  for (let i = 0; i < cycleCount; ++i) {
+  // If there is only one cycle in the array we will not make a prediction more than one cycle. Because we don't know the exact length of the cycle
+  if (cycles.length === 1) return forecastDates;
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:436]: `{` `}`

File: src/state/CalculationLogics.ts [hunk 9]
```diff
@@ -428,70 +456,73 @@ export function isPeriodToday(cycles: Cycle[]) {
   return dayOfCycle <= cycles[0].periodLength;
 }
 
-export function getOvulationDays(cycles: Cycle[]) {
-  if (cycles.length < 2) {
-    return [];
-  }
+export function getOvulationDates(cycles: Cycle[]) {
+  if (cycles.length < 2) return [];
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:460]: `{` `}`

File: src/state/CalculationLogics.ts [hunk 10]
```diff
@@ -276,20 +292,21 @@ export function getNewCyclesHistory(periodDays: string[]) {
...
     const date = startOfDay(new Date(periodDays[i]));
     const prevDate = startOfDay(new Date(periodDays[i - 1]));
     const diffInDays = differenceInDays(date, prevDate);
 
-    if (diffInDays <= 1) {
-      newCycles[0].periodLength++;
-    } else if (diffInDays <= 2) {
-      newCycles[0].periodLength += 2;
+    if (diffInDays <= 2) {
+      newCycles[0].periodLength += diffInDays === 1 ? 1 : 2;
```
Comment (imblowfish) [inline src/state/CalculationLogics.ts:309]: Unfortunately I don't understand what's going on here

If `diffInDays` here equals to `0`, `-1` or lesser why we adding `2` here?

The other question if `diffInDays` may be `1` or `2` than we you don't needed in this check here. Instead you just can write `newCycles[0].periodLength += diffInDays`

**Review by imblowfish — CHANGES_REQUESTED** [review]

File: src/state/CalculationLogics.ts (hunk 10 above)
Comment (IraSoro) [reply src/state/CalculationLogics.ts]: You're right, I fixed it.

**Review by imblowfish — APPROVED** [review]

**Pull request merged**